holoviews/plotting/util.py

```diff
@@ -82,55 +82,56 @@ def isoverlay_fn(obj):
 
 def compute_overlayable_zorders(obj, path=[]):
     """
-    Traverses the Callables on DynamicMap to determine which objects
-    in the graph are associated with specific (Nd)Overlay layers.
-    Returns a mapping between the zorders of each layer and lists of
-    objects associated with each.
+    Traverses an overlayable composite container to determine which
+    objects are associated with specific (Nd)Overlay layers by
+    z-order, making sure to take DynamicMap Callables into
+    account. Returns a mapping between the zorders of each layer and a
+    corresponding lists of objects.
 
-    Used to determine which subplots should be linked with Stream
-    callbacks.
+    Used to determine which overlaid subplots should be linked with
+    Stream callbacks.
     """
     path = path+[obj]
     zorder_map = defaultdict(list)
 
     # Process non-dynamic layers
     if not isinstance(obj, DynamicMap):
         if isinstance(obj, CompositeOverlay):
-            for i, o in enumerate(obj):
-                zorder_map[i] = [o, obj]
+            for z, o in enumerate(obj):
+                zorder_map[z] = [o, obj]
         else:
             if obj not in zorder_map[0]:
                 zorder_map[0].append(obj)
         return zorder_map
 
     isoverlay = isinstance(obj.last, CompositeOverlay)
     isdynoverlay = obj.callback._is_overlay
-    found = any(isinstance(p, DynamicMap) and p.callback._overlay for p in path)
     if obj not in zorder_map[0] and not isoverlay:
         zorder_map[0].append(obj)
 
     # Process the inputs of the DynamicMap callback
     dmap_inputs = obj.callback.inputs if obj.callback.link_inputs else []
-    for i, inp in enumerate(dmap_inputs):
+    for z, inp in enumerate(dmap_inputs):
         if any(not (isoverlay_fn(p) or p.last is None) for p in path) and isoverlay_fn(inp):
             # Skips branches of graph that collapse Overlay layers
             # to avoid adding layers that have been reduced or removed
             continue
 
         # Recurse into DynamicMap.callback.inputs and update zorder_map
```

Comment:
>So this would not happen for a DynamicMap callback with inputs that is user defined i.e isdynoverlay is False because _is_overlay wouldn't be set to True by the user? 

That's correct, the branch below handles that by iterating over the items in the user defined DynamicMap, which will ensure the zorder is maintained but won't attempt to figure out linked inputs.